gcloud/storage/batch.py

```diff
@@ -0,0 +1,208 @@
+# Copyright 2014 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Batch updates / deletes of storage buckets / blobs.
+
+See: https://cloud.google.com/storage/docs/json_api/v1/how-tos/batch
+"""
+from email.encoders import encode_noop
+from email.generator import Generator
+from email.mime.application import MIMEApplication
+from email.mime.multipart import MIMEMultipart
+from email.parser import Parser
+import io
+import json
+import sys
+
+import six
+
+from gcloud._localstack import _LocalStack
+
+
+_BATCHES = _LocalStack()
+
+_PROXIED_ATTRS = [
```

Comment:
The `Blob` stuff is a concern but I'm not sure we can even use upload in a `batch`.

The only other place `.http` is used is in `Connection.api_request`:

```
$ # Current HEAD
$ git log -1 --pretty=%H
56f60bb94b313c397f5aeadee1f0e1f6e5b5290c
$ git grep -n '\.http' -- gcloud/storage/ | egrep -v test
gcloud/storage/blob.py:236:        download.InitializeDownload(request, self.connection.http)
gcloud/storage/blob.py:347:        upload.InitializeUpload(request, conn.http)
gcloud/storage/blob.py:356:            http_wrapper.MakeRequest(conn.http, request, retries=num_retries)
gcloud/storage/connection.py:169:        return self.http.request(uri=url, method=method, headers=headers,
```

Also, per our previous [discussion][1] we discussed getting rid of essentially everything except for `api_request`.

It seems we could make batches work in much the same way as in `datastore`, but `storage` needs a redesign (a la #632). This would be much easier if `Bucket` didn't own a connection (and other objects indirectly via ownership of a bucket).

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/pull/604#issuecomment-73929760